pkg/frontend/nvme.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/ulule/deepcopier"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+const (
+	mevTransportType = "npi"
+)
+
+// CreateNVMeController creates an NVMe Controller
+func (s *Server) CreateNVMeController(ctx context.Context, in *pb.CreateNVMeControllerRequest) (*pb.NVMeController, error) {
+	log.Printf("CreateNVMeController: Received from client: %v", in)
+
+	err := verifyCreateNvmeControllerRequestArgs(in)
+	if err != nil {
+		return nil, err
+	}
+
+	subsys, ok := s.nvme.Subsystems[in.NvMeController.Spec.SubsystemId.Value]
+	if !ok {
+		msg := fmt.Sprintf("unable to find subsystem with id %v", in.NvMeController.Spec.SubsystemId.Value)
+		log.Printf("error: %v", msg)
+		return nil, status.Error(codes.FailedPrecondition, msg)
+	}
+	log.Printf("Found subsystem: %v", subsys.Spec.Nqn)
+
+	controllerID := in.NvMeController.Spec.Id.Value
+	_, ok = s.nvme.Controllers[controllerID]
+	if ok {
+		msg := fmt.Sprintf("Already existing controller with id %v", controllerID)
+		log.Printf("error: %v", msg)
+		return nil, status.Error(codes.AlreadyExists, msg)
+	}
+
+	params := models.NvmfSubsystemAddListenerParams{}
+	params.Nqn = subsys.Spec.Nqn
+	params.ListenAddress.Trtype = mevTransportType
+	params.ListenAddress.Traddr = calculateTransportAddr(in.NvMeController.Spec.PcieId)
+
+	var result models.NvmfSubsystemAddListenerResult
+	err = s.rpc.Call("nvmf_subsystem_add_listener", &params, &result)
+	if err != nil {
+		log.Printf("error: %v", err)
+		return nil, status.Errorf(codes.FailedPrecondition, "failed to execute SPDK call")
+	}
+	log.Printf("Received from SPDK: %v", result)
+	if !result {
+		msg := fmt.Sprintf("could not create NVMe controller: %v", controllerID)
+		log.Print(msg)
+		return nil, status.Errorf(codes.FailedPrecondition, msg)
+	}
+
+	in.NvMeController.Spec.NvmeControllerId = -1
+	in.NvMeController.Status = &pb.NVMeControllerStatus{Active: true}
+	s.nvme.Controllers[controllerID] = in.NvMeController
+
+	response := &pb.NVMeController{}
+	err = deepcopier.Copy(in.NvMeController).To(response)
+	if err != nil {
+		log.Printf("error: %v", err)
+		return nil, status.Errorf(codes.Internal, "Failed to create response")
+	}
+	return response, nil
+}
+
+// DeleteNVMeController deletes an NVMe Controller
+func (s *Server) DeleteNVMeController(ctx context.Context, in *pb.DeleteNVMeControllerRequest) (*emptypb.Empty, error) {
+	log.Printf("DeleteNVMeController: Received from client: %v", in)
+	controller, ok := s.nvme.Controllers[in.Name]
+	if !ok {
+		msg := fmt.Sprintf("unable to find controller with id %v", in.Name)
+		log.Printf("error: %v", msg)
+		return nil, status.Error(codes.FailedPrecondition, msg)
+	}
+
+	subsys, ok := s.nvme.Subsystems[controller.Spec.SubsystemId.Value]
+	if !ok {
+		msg := fmt.Sprintf("unable to find subsystem with id %v for controller %v",
+			controller.Spec.SubsystemId.Value,
+			in.Name)
+		log.Printf("error: %v", msg)
```

Comment:
same comment here about idempotency: why instead of error, not to return the success here if controller already deleted ?